community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxIteratorRelationshipCursor.java

```
@@ -0,0 +1,98 @@
+import org.neo4j.kernel.api.txstate.TransactionState;
+import org.neo4j.kernel.impl.api.store.RelationshipIterator;
+
+/**
+ * Overlays transaction state on a {@link RelationshipCursor}. This additionally knows how to traverse added
+ * relationships in this transaction and skip deleted ones.
+ */
+public class TxIteratorRelationshipCursor extends TxAbstractRelationshipCursor
+{
+    private RelationshipIterator added;
+    private RelationshipIterator addedRelationshipIterator;
+
+    public TxIteratorRelationshipCursor( TransactionState state, Consumer<TxIteratorRelationshipCursor> instanceCache )
+    {
+        super( state, (Consumer) instanceCache );
+    }
+
+    public TxIteratorRelationshipCursor init( RelationshipCursor cursor,
+            RelationshipIterator addedRelationshipIterator )
+    {
+        super.init( cursor );
+
+        this.relationshipIsAddedInThisTx = false;
+        this.addedRelationshipIterator = addedRelationshipIterator;
+
+        added = null;
+
+        return this;
+    }
+
+    @Override
+    public boolean next()
+    {
+        if ( added == null )
+        {
+            while ( cursor.next() )
+            {
+                long id = cursor.getId();
+
+                if ( state.relationshipIsDeletedInThisTx( id ) )
+                {
+                    continue;
+                }
+
+                visit( id, cursor.getType(), cursor.getStartNode(), cursor.getEndNode() );
+                relationshipState = state.getRelationshipState( cursor.getId() );
+                return true;
+            }
+
+            added = addedRelationshipIterator;
+            relationshipIsAddedInThisTx = true;
+        }
+
+        if ( added.hasNext() )
+        {
+            added.relationshipVisit( added.next(), this );
+            relationshipState = state.getRelationshipState( cursor.getId() );
+            return true;
+        }
+        else
+        {
+            visit( -1, -1, -1, -1 );
```

Restating my concern here about failing early: to me it feels like calling an accessor method after a call to `next` that returns `false` should throw exception, not returning `-1` or similar. I think throwing is fine because that helps fleshing out bad usages of the cursors earlier, i.e. trivial bugs in the code.